tutorials/snapcraft/snap-a-nodejs-app.md

```diff
@@ -0,0 +1,666 @@
+
+
+
+```bash
+$ sudo chuck-norris-webserver.node-service 
+Server listening on: http://localhost:80
+```
+
+You can now point your browser to [http://localhost] and enjoy some Chuck Norris quotes! All debug outputs will be visible on the command line prompt. Press Ctrl+C to exit it.
+Having a command while you are working on your snap helps you greatly in inspecting what’s going on in your snap. This is why we strongly recommend that approach.
+
+Also, that means to listen to that port 80, we’ll need the service to run as a privileged user. The good news is that in the snap world, thanks to confinement, all services are running as root while being safe!
+It’s something to keep in mind while you iterate on a command which is going to be turned into a service later on: run them as root to reproduce the exact running context!
+
+It’s now high time to get rid of this devmode thing and add some confinement to your snap, let’s do this right away!
+
+ 
+## Turning on confinement!
+Duration: 5:00
+
+
+```bash
+Lost or starting from here?
+Check or download [here-3] to see what your current directory should look like.
+```
+
+### Enabling confinement
+
+In our first snap tutorial, we simply turned confinement on without any additional work. Let’s try this as well!
+You don’t forcefully need to change your `snapcraft.yaml` and rebuild (we’ll do it later), both `snap try` and `snap install` have an option to force confinement on snaps declaring devmode.
+
+
+```bash
+$ snap try prime/ --jailmode
+chuck-norris-webserver 1.0.0 mounted from […]/prime
+$ sudo chuck-norris-webserver.node-service 
+events.js:141
+      throw er; // Unhandled 'error' event
+      ^
+
+Error: listen EACCES 0.0.0.0:80
+    at Object.exports._errnoException (util.js:870:11)
+    at exports._exceptionWithHostPort (util.js:893:20)
+    at Server._listen2 (net.js:1221:19)
+    at listen (net.js:1270:10)
+    at Server.listen (net.js:1366:5)
+    at Object.<anonymous> (/snap/chuck-norris-webserver/x2/lib/node_modules/chuck-norris-app/main.js:71:8)
+    at Module._compile (module.js:409:26)
+    at Object.Module._extensions..js (module.js:416:10)
+    at Module.load (module.js:343:32)
+    at Function.Module._load (module.js:300:12)
+```
+
+
+Thanks to keeping the service as a simple command, we see easily that, even if ran as root, the nodejs built-in web server isn’t able to listen on port 80.
+
+### What do we need to do?
+
+The snap not being in devmode anymore, it needs to get additional permissions. Those are given through an interface. We thus have to find which interface our snap will need, and declare a plug to that interface. The snap plug will connect to the core slot which declares this interface. For an easy analogy, thinks of a plug like an electrical plug, slot is what you put the plug into, and the interface is the kind of data (permissions here) which will be send over via that interface.
```

Comment:
Occurences of "Interface", "plug" and "slot" are bold in the source:

"The snap not being in devmode anymore, it needs to get additional permissions. Those are given through an **interface**. We thus have to find which **interface** our snap will need, and declare a **plug** to that interface. The snap **plug** will connect to the core **slot** which declares this **interface**. For an easy analogy, thinks of a plug like an electrical plug, slot is what you put the plug into, and the interface is the kind of data (permissions here) which will be send over via that interface."